services/CurrentTime/include/ble-service-current-time/CurrentTimeService.h

```diff
@@ -0,0 +1,186 @@
+
+#ifndef CURRENT_TIME_SERVICE_H
+#define CURRENT_TIME_SERVICE_H
+
+#ifdef BLE_FEATURE_GATT_SERVER
+
+#include "ble/BLE.h"
+#include "ble/Gap.h"
+#include "events/EventQueue.h"
+#include "mbed_rtc_time.h"
+#include "Timer.h"
+
+#include <ctime>
+
+/**
+ * Current Time Service
+ *
+ * @par purpose
+ * The current time service allows a BLE device to expose date and time information to other BLE devices.
+ *
+ * @par usage
+ * The on_current_time_changed() event handler should be overridden by your application
+ *
+ * @note The specification for the current time service can be found here:
+ * https://www.bluetooth.com/specifications/gatt
+ *
+ * @attention The user should not instantiate more than a single current time service service
+ */
+class CurrentTimeService {
+public:
+    static const uint8_t MANUAL_TIME_UPDATE             = 1 << 0;
+    static const uint8_t EXTERNAL_REFERENCE_TIME_UPDATE = 1 << 1;
+    static const uint8_t CHANGE_OF_TIME_ZONE            = 1 << 2;
+    static const uint8_t CHANGE_OF_DST                  = 1 << 3;
+
+    static constexpr std::chrono::seconds UPDATE_TIME_PERIOD = std::chrono::seconds(60);
+
+    struct EventHandler {
+        /**
+         * This function is called if the current time characteristic is changed by the client
+         */
+        virtual void on_current_time_changed(time_t current_time, uint8_t adjust_reason) { }
+    };
+
+    /**
+     * Initialize the internal BLE object to @p ble and configure the current time characteristic
+     * with the appropriate UUID.
+     *
+     * @param ble BLE object to host the current time service
+     *
+     * @attention The Initializer must be called after instantiating a current time service.
+     */
+    CurrentTimeService(BLE &ble, events::EventQueue &event_queue);
+
+    ~CurrentTimeService() = default;
+
+    CurrentTimeService(const CurrentTimeService&) = delete;
+    CurrentTimeService &operator=(const CurrentTimeService&) = delete;
+
+    /**
+     * Set the onCurrentTimeRead() and onCurrentTimeWritten() functions as the read authorization callback
+     * and write authorization callback, respectively for the current time characteristic.
+     * Add the current time service to the BLE device.
+     *
+     * @return BLE_ERROR_NONE if the service was successfully added.
+     */
+    ble_error_t init();
+
+    /**
+    * Set the event handler to handle events raised by the current time service.
+    *
+    * @param handler EventHandler object.
+    */
+    void set_event_handler(EventHandler *handler);
+
+    /**
+     * Get the time in seconds since 00:00 January 1, 1970 plus a configurable offset.
+     *
+     * @return Time in seconds.
+     */
+    time_t get_time() const;
+
+    /**
+     * Set the time offset, i.e. the time in seconds beyond Epoch time.
+     *
+     * @param host_time Time in seconds according to your host.
+     * @param adjust_reason Bitmask using a combination of MANUAL_TIME_UPDATE, EXTERNAL_REFERENCE_TIME_UPDATE,
```

Comment:
```suggestion
     * @param adjust_reason Bitmask using one or a combination of MANUAL_TIME_UPDATE, EXTERNAL_REFERENCE_TIME_UPDATE,
```

Reply:
A combination can exclude all but one?

Reply:
good point, technically you can have an unknown reason
```suggestion
     * @param adjust_reason Bitmask of one of or a combination of MANUAL_TIME_UPDATE, EXTERNAL_REFERENCE_TIME_UPDATE or 0 if reason is unknown,
```